analyzers/src/SonarAnalyzer.CSharp/Rules/ArrayPassedAsParams.cs

```diff
@@ -0,0 +1,51 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ArrayPassedAsParams : ArrayPassedAsParamsBase<SyntaxKind>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+    protected override bool ShouldReport(SonarSyntaxNodeReportingContext context, SyntaxNode expression) =>
+        expression switch
+        {
+            ObjectCreationExpressionSyntax { } creation =>
+                CheckLastArgument(creation.ArgumentList) && IsParamParameter(context, creation, creation.ArgumentList.Arguments.Last()),
+            InvocationExpressionSyntax { } invocation =>
+                CheckLastArgument(invocation.ArgumentList) && IsParamParameter(context, invocation, invocation.ArgumentList.Arguments.Last()),
+            _ => false
+        };
+
+    protected override Location GetLocation(SyntaxNode expression) =>
+        expression switch
+        {
+            ObjectCreationExpressionSyntax { } creation => creation.ArgumentList.Arguments.Last().Expression.GetLocation(),
+            InvocationExpressionSyntax { } invocation => invocation.ArgumentList.Arguments.Last().Expression.GetLocation(),
+            _ => expression.GetLocation()
+        };
+
+    private static bool CheckLastArgument(ArgumentListSyntax argumentList) =>
+        argumentList is not null
+        && argumentList.Arguments.Any()
+        && argumentList.Arguments.Last().Expression is ArrayCreationExpressionSyntax invocationArray
+        && invocationArray.Initializer is InitializerExpressionSyntax { Expressions.Count: > 0 };
```

Comment:
Now let's move these to the base class
* We know, that the node we register for will be passed to the `invocation` parameter of `IsParamParameter`. This is language-independent.
* We need to extract the last argument from the argument list but only if it is an array creation. Let's keep this language dependent.
* But we need to report on the last argument. That part can be made language-independent:
Overall the final solution should look about like this for the base class:
```cs
public abstract class ArrayPassedAsParamsBase<TSyntaxKind, TArgumentNode> : SonarDiagnosticAnalyzer<TSyntaxKind>
    where TSyntaxKind : struct
    where TArgumentNode : SyntaxNode
{
    protected abstract TSyntaxKind[] ParamsInvocationKinds { get; }
    // Returns the last argument, but only if the last argument is an array creation expression with at least one element.
    protected abstract TArgumentNode GetLastArgumentIfArrayCreation(SyntaxNode invocation);

    protected sealed override void Initialize(SonarAnalysisContext context)
    {
        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, CheckExpression, ParamsInvocationKinds);
    }

    private void CheckExpression(SonarSyntaxNodeReportingContext context)
    {
        if (GetLastArgumentIfArrayCreation(context.Node) is lastArgument
            && IsParamParameter(context, context.Node, lastAgrument)
        {
            Report(context, lastArgument.GetLocation());
        }
    }
}
```

Reply:
I'm amazed by your reorganizing skill 💯 Thanks for the lesson!